cas-server-core/src/main/java/org/jasig/cas/authentication/principal/cache/AbstractPrincipalAttributesRepository.java

```diff
@@ -0,0 +1,271 @@
+     * Convert principal attributes to person attributes.
+     * @param p  the principal carrying attributes
+     * @return person attributes
+     */
+    private Map<String, List<Object>> convertPrincipalAttributesToPersonAttributes(final Principal p) {
+        final Map<String, List<Object>> convertedAttributes = new HashMap<>(p.getAttributes().size());
+        final Map<String, Object> principalAttributes = p.getAttributes();
+
+        for (final Map.Entry<String, Object> entry : principalAttributes.entrySet()) {
+            final Object values = entry.getValue();
+            final String key = entry.getKey();
+            if (values instanceof List) {
+                convertedAttributes.put(key, (List) values);
+            } else {
+                convertedAttributes.put(key, Collections.singletonList(values));
+            }
+        }
+        return convertedAttributes;
+    }
+
+    /**
+     * Obtains attributes first from the repository by calling
+     * {@link org.jasig.services.persondir.IPersonAttributeDao#getPerson(String)}.
+     *
+     * @param id the person id to locate in the attribute repository
+     * @return the map of attributes
+     */
+    protected final Map<String, List<Object>> retrievePersonAttributesToPrincipalAttributes(final String id) {
+
+        final IPersonAttributes personAttributes = this.attributeRepository.getPerson(id);
+        if (personAttributes == null) {
+            logger.debug("Could not find principal [{}] in the repository so no attributes are returned.", id);
+            return Collections.emptyMap();
+        }
+
+        final Map<String, List<Object>> attributes = personAttributes.getAttributes();
+        if (attributes == null) {
+            logger.debug("Principal [{}] has no attributes and so none are returned.", id);
+            return Collections.emptyMap();
+        }
+        return attributes;
+    }
+
+    @Override
+    public final Map<String, Object> getAttributes(@NotNull final Principal p) {
+        final Map<String, Object> cachedAttributes = getPrincipalAttributesFromCache(p);
+        if (cachedAttributes != null && !cachedAttributes.isEmpty()) {
+            logger.debug("Found [{}] cached attributes for principal [{}]", cachedAttributes.size(), p.getId());
+            return cachedAttributes;
+        }
+
+        final Map<String, List<Object>> sourceAttributes = retrievePersonAttributesToPrincipalAttributes(p.getId());
+        logger.debug("Found [{}] attributes for principal [{}] from the attribute repository.",
+                sourceAttributes.size(), p.getId());
+
+        if (this.mergingStrategy == null) {
+            logger.debug("No merging strategy found, so attributes retrieved from the repository will be used instead.");
+            return convertAttributesToPrincipalAttributesAndCache(p, sourceAttributes);
+        }
+
+        final Map<String, List<Object>> principalAttributes = convertPrincipalAttributesToPersonAttributes(p);
```

Comment:
So the process is complicated, but needed:

1. First person attributes are retrieved. 
2. If there is no merging required, simply convert to principal and return. 90% of cases.
3. If merging is required, merge principal attributes with person attributes, and reconvert, and return. The step you refer to as not needed is used for merging to resolve collisions.